lib/elixir/lib/kernel/typespec.ex

```diff
@@ -192,20 +192,65 @@ defmodule Kernel.Typespec do
 
   @doc false
   def translate_typespecs_for_module(_set, bag) do
-    types = Enum.map(take_typespec(bag, :type), &translate_type/1)
-    specs = Enum.map(take_typespec(bag, :spec), &translate_spec/1)
-    callbacks = Enum.map(take_typespec(bag, :callback), &translate_spec/1)
-    macrocallbacks = Enum.map(take_typespec(bag, :macrocallback), &translate_spec/1)
+    type_typespecs = take_typespec(bag, :type)
+    defined_type_pairs = collect_defined_type_pairs(type_typespecs)
+    state = %{defined_type_pairs: defined_type_pairs, used_type_pairs: []}
+
+    {types, state} = Enum.map_reduce(type_typespecs, state, &translate_type/2)
+    {specs, state} = Enum.map_reduce(take_typespec(bag, :spec), state, &translate_spec/2)
+    {callbacks, state} = Enum.map_reduce(take_typespec(bag, :callback), state, &translate_spec/2)
+
+    {macrocallbacks, state} =
+      Enum.map_reduce(take_typespec(bag, :macrocallback), state, &translate_spec/2)
+
     optional_callbacks = List.flatten(get_typespec(bag, {:accumulate, :optional_callbacks}))
-    {types, specs, callbacks, macrocallbacks, optional_callbacks}
+    used_types = filter_used_types(types, state)
+
+    {used_types, specs, callbacks, macrocallbacks, optional_callbacks}
   end
 
   defp take_typespec(bag, key) do
     :ets.take(bag, key)
   end
 
-  defp translate_type({_, {kind, {:::, _, [{name, _, args}, definition]}, pos}})
-       when is_atom(name) and name != ::: do
+  defp collect_defined_type_pairs(type_typespecs) do
+    Enum.reduce(type_typespecs, %{}, fn {_, {_, expr, pos}}, type_pairs ->
+      env = :elixir_locals.get_cached_env(pos)
+
+      case type_to_signature(expr) do
+        {name, arity} = type_pair ->
+          if builtin_type?(name, arity) do
+            message = "type #{name}/#{arity} is a builtin type and it cannot be redefined"
+            compile_error(env, message)
+          end
+
+          if Map.has_key?(type_pairs, type_pair) do
+            compile_error(env, "type #{name}/#{arity} is already defined")
+          end
+
+          Map.put(type_pairs, type_pair, env)
```

Comment:
This is only used in `filter_used_types/2`, so I might store `{file, line}` as value instead of the whole `env`, wdyt?

Reply:
:+1: